### ISSUE TITLE
Compiler sometimes ignores declared type of variables

### ISSUE THREAD

**Issue by johnw42** [microsoft/TypeScript#55516]
### 🔎 Search Terms

type annotation ignore; type annotation respect; declared type ignore; declared type respect

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about type system behavior

### ⏯ Playground Link

https://www.typescriptlang.org/play?strictFunctionTypes=false&ts=5.3.0-dev.20230825&q=454#code/JYOwLgpgTgZghgYwgAgCpQig3gKGchAC2ABsATDEALjQwgG0BdAbhwF8ccYBXEBMYAHsQyEoMEBrAKoAHAMLFyACjB0a6TABpkoMsCQBnGiG4BbAEbQmASnV1kAH2S8yEGKAhlkufCQhhkQmgIO0xHZxBXdxBPZABeZFVMVnwYQShkJQRhAwDdCAAPZEEYHUj9CANrbzx8ZAB6euQAZUJBbnJkDBl0gOgodMDg5EsEOG4DFAByIlIKCBApnQNkEEEAuGRTCAtoYtKXNw8yADpa-CCMeKGME9nySnp8gpZajnwMMG4oEUuIVg4OEaaAAnjIIABRKADKA0ORwEBrAIYOBeGQDcFQASVfYRKLHTIovQgADmyBminmi2sODEklkCjmSiwBEplBoTGQbG09AADNpeYxrMwgA

### 💻 Code

```ts
interface Tree {
  children: Tree[];
}

function lookUpChild(tree: Tree, indices: number[]): Tree | undefined {
  let here: Tree | undefined = tree;
  for (const index of indices) {
    // Should report error here because 'children' is not a member of undefined.
    here = here.children[index];
  }
  return here;
}

// TypeError: Cannot read properties of undefined (reading 'children')
lookUpChild({ children: [] }, [0, 0]);
```

### 🙁 Actual behavior

The type of `here` is `Tree`, despite the type annotation. This can be verified by inspecting the type of `here` in the playground. The compiler reports no error for code that causes a runtime error when executed.

### 🙂 Expected behavior

The type of `here` should be `Tree | undefined`. The indicated line should have a type error.

### Additional information about the issue

The loop in the code isn't necessary to reproduce the problem. I included it mainly to make an example that actually throws an error. It was also part of the original code where I discovered the problem.

Using the `at` method on the array instead of subscripting causes the correct type error to be reported. That seems to demonstrate that the problem comes from overly aggressive type narrowing.

**Comment by fatcerberus** [comment]
The variable is narrowed to `Tree` on assignment; that part is intentional.  The problem is that array access is assumed to be in-bounds; you can try enabling `noUncheckedIndexedAccess` but it might be more trouble than it’s worth (which is why it’s not enabled by default).
